### DISCUSSION TITLE
I need help understanding and configuring a multi monitor setup.

### DISCUSSION THREAD

**Discussion by r-gorajski** [qtile/qtile#3269] [Q&A]
Please help me understand and configure the multi monitor setup.

At the moment I am at the crossroads between i3wm and Qtile. I like i3 very much, but Qtile is more flexible in my opinion. Besides, it's time for something new. Despite my poor knowledge of Python, the configuration that satisfies me so far didn't take long. Unfortunately, I completely don't understand how Qtile treats multiple monitors. Currently on i3, I have configured by default - workspace 2,4,5,6 on the left HDMI1 monitor, the rest on the right DP1. The shift of focus between workspaces is Mod4 + 1 to 0 (10 sec). When I'm on ws: 1 and I press mod4 + 2, the focus is transferred to the left HDMI1 monitor and that's it. Qtile does something strange, swaps monitors? I totally don't get it.

I know my question might be hard for someone who hasn't used the i3 before. But maybe some of you will take pity on me and explain how it works in Qtile and suggest how to change the Qtile configuration to work like i3 when it comes to multi monitors configuration?

I found a few people on the internet with a similar problem in this regard, but i don't know how to solve it.

Thank you very much in advance for your help.

Regards Robert

**Comment by elParaguayo** [comment]
Welcome to Qtile! Glad to hear you're liking it so far.

By default, groups aren't attached to screens so, when you select a group, it is shown on the current screen.

However, if you want to lock a group to a particular screen, it is possible. Have a look here: http://docs.qtile.org/en/stable/manual/faq.html#how-can-i-get-my-groups-to-stick-to-screens

If you still need some help after that then just let us know.

**Comment by ExpandingMan** [comment]
I'm also having a hell of a time figuring out what to do with multiple screens.  The FAQ is helpful in that it explains a very specific use case, but I'm having a much harder time understanding how the API works.

For one, I can't for the life of me figure out how to access the [top level `qtile` object](https://docs.qtile.org/en/stable/manual/ref/commands.html?highlight=focus#qtile) that has commands like `cmd_next_screen()`.

**Reply by elParaguayo** [reply]
Yeah. There are definitely some gaps in the docs relating to available commands. A fix is in the works...

As for your particular query, top level `cmd_` commands can be accessed via `lazy.next_screen()`. NB the `lazy` calls only work as keybindings or widget mouse callbacks.

**Reply by ExpandingMan** [reply]
Thanks, that's helpful...

Are all the commands listed in the documentation?  For example, I don't see any commands listed [under groups](https://docs.qtile.org/en/stable/manual/ref/commands.html#group) so I'm not sure where to find those.

**Reply by ExpandingMan** [reply]
Not sure how up to date it is but [this config example](https://github.com/justinesmithies/qtile-x-dotfiles/blob/master/.config/qtile/config.py) has been the most useful for me by far up to this point.

**Reply by elParaguayo** [reply]
> Are all the commands listed in the documentation? For example, I don't see any commands listed [under groups](https://docs.qtile.org/en/stable/manual/ref/commands.html#group) so I'm not sure where to find those.

No. They're not. I've got a PR that will ensure everything is listed.

**Comment by ExpandingMan** [comment]
Ok, so what I'm trying to do now is that I want to be able to focus any group by jumping to the screen that group is currently in.  Then I want a separate keybinding to go to move the group between screens.

I tried several variations on
```python
@lazy.function
def togroup(qtile, g):
    qtile.to_screen(g.screen.info()["index"])
    g.toscreen()
```
for the former but so far nothing works.

**Reply by elParaguayo** [reply]
Just to check - you want to be able to focus a group but have it focus on the screen where the group is currently located, rather than have the group move to the current screen?

**Reply by ExpandingMan** [reply]
Yes, that's right.  I think the fundamental problem that I'm having is that the `qtile` object I'm able to access is not the same as whatever it is I'm sending commands to with my `InteractiveCommandClient()` REPL, so I'm having a hard time figuring out how to get objects out of it.

**Reply by elParaguayo** [reply]
How are you binding that command in your config?

**Reply by ExpandingMan** [reply]
here's an example of what I'm trying to do
```python
@lazy.function
def togroup(qtile, g):
    g = qtile.groups[g.name]
    cmd.to_screen(g.screen.info()["index"])
    g.toscreen()

def keymap_group(g):
    return [
            # mod1 + letter of group = switch to group
            Key(
                [mod],
                g.name,
                #lazy.group[g.name].toscreen(),
                togroup(g),
                desc=f"Switch to group {g.name}",
                ),
            # mod1 + shift + letter of group = switch to & move focused window to group
            Key(
                [mod, "shift"],
                g.name,
                lazy.window.togroup(g.name, switch_group=False),
                desc=f"Switch to & move focused window to group {g.name}",
                ),
            ]
```

**Reply by elParaguayo** [reply]
Can you try this:
``` python
@lazy.function
def togroup(qtile, g):
    qtile.cmd_to_screen(g.screen.info()["screen"])
    g.cmd_toscreen()
```

**Reply by ExpandingMan** [reply]
I had tried something like that before, I get
```python
 124   │ Traceback (most recent call last):
 125   │   File "/usr/lib/python3.10/site-packages/libqtile/command/base.py", line 207, in cmd_function
 126   │     function(self, *args, **kwargs)
 127   │   File "/home/expandingman/.config/qtile/config.py", line 54, in togroup
 128   │     qtile.cmd_to_screen(g.screen.info()["screen"])
 129   │ AttributeError: 'Group' object has no attribute 'screen'
```
It appears the object `Group` that's created in the config is completely different than those returned by the command interface.

**Reply by elParaguayo** [reply]
Sorry `g.info()["screen"]`

**Reply by ExpandingMan** [reply]
Similar
```python
 148   │ Traceback (most recent call last):
 149   │   File "/usr/lib/python3.10/site-packages/libqtile/command/base.py", line 207, in cmd_function
 150   │     function(self, *args, **kwargs)
 151   │   File "/home/expandingman/.config/qtile/config.py", line 54, in togroup
 152   │     qtile.cmd_to_screen(g.info()["screen"])
 153   │ AttributeError: 'Group' object has no attribute 'info'
 ```

**Reply by elParaguayo** [reply]
Sorry - am rushing things. I need to finish something else first and then I'll post a proper solution for you.

**Reply by elParaguayo** [reply]
Also, what do you want to happen if the group isn't on a screen?

**Reply by ExpandingMan** [reply]
Not on *any* screen?  I hadn't thought of it, I can't get myself into that state in `i3`.  For the time being I guess it would be fine if it just defaulted to screen 0.

More broadly, despite having read much of the [commands API documentation](https://docs.qtile.org/en/stable/manual/commands/index.html) I'm *very* confused about how to interact with qtile in the configuration file.  As far as I can tell there are several similar but distinct overlapping API's and I can't seem to get them straight.  The `Group`s that I create in the config are not the same as what I can get back from `qtile`, and none of what I can access with `qtile` seems to match the API that's available through `CommandInterface`.  Also, for reasons I can't quite understand, `CommandInterface` doesn't seem to work form the config.

I can probably figure out all of this screen stuff on my own, but first I need to figure out what is going on with the API.  The `CommandInterface` I can access through the REPL more or less makes sense to me, but so far my attempts to treat the config the same way have resulted in total chaos.

**Reply by elParaguayo** [reply]
Try something like this
``` python
@lazy.function
def togroup(qtile, g):
    # Get the group object
    group = qtile.groups_map[g.name]
    # Get the screen the group is on (fall back to 0)
    screen = group.info()["screen"] or 0
    # Focus the screen
    qtile.cmd_to_screen(screen)
    # Move group to that screen
    group.cmd_toscreen()
```

**Reply by elParaguayo** [reply]
> More broadly, despite having read much of the [commands API documentation](https://docs.qtile.org/en/stable/manual/commands/index.html) I'm _very_ confused about how to interact with qtile in the configuration file. As far as I can tell there are several similar but distinct overlapping API's and I can't seem to get them straight. The `Group`s that I create in the config are not the same as what I can get back from `qtile`, and none of what I can access with `qtile` seems to match the API that's available through `CommandInterface`. Also, for reasons I can't quite understand, `CommandInterface` doesn't seem to work form the config.

Yes. This is a recurring issue which I'm trying to simplify a bit. A lot of the confusion comes from the fact that the command names used via the command interface look different to the internal method names. This is because, currently, the code uses the `cmd_` prefix to expose commands over the interface and removes the prefix at that point.

You are also correct that the config `Group` is not the same as the `Group` objects exposed via the interface,

You shouldn't really use `CommandInterface` in your config. You can use `lazy` commands (which are a type of `CommandClient`) in keybindings and mouse callbacks. If you want to do other things then you can access them via the `qtile` object but you just have to remember to use the full method names.

#2727 will remove some of the confusion as it removes the `cmd_` prefix so internal methods will have the same name as exposed commands.

> 
> I can probably figure out all of this screen stuff on my own, but first I need to figure out what is going on with the API. The `CommandInterface` I can access through the REPL more or less makes sense to me, but so far my attempts to treat the config the same way have resulted in total chaos.

If you're using `CommandInterface` in a REPL session then these commands will be like those in the `lazy` keybindings.

**Reply by ExpandingMan** [reply]
Thanks for all your help!

This clears things up somewhat, though I look forward to the API getting less confusing.

Funnily enough, I was experimenting a bit, and despite some initial resistance because it was so unfamiliar, I've ultimately discovered that qtile's default behavior is *way better* than what I was trying to do here, provided you can find good keybindings for swapping screen focus.  So ironically I wound up sticking with the default anyway.

**Reply by elParaguayo** [reply]
To swap screen focus you can use `lazy.next_screen()`, `lazy.prev_screen()` and `lazy.to_screen(0)`.

**Reply by ExpandingMan** [reply]
Thanks.  Is there a way I can query what is the current screen?

**Reply by elParaguayo** [reply]
In a function? `qtile.current_screen`.

**Reply by ExpandingMan** [reply]
Hm... so something really weird happens... if I have it default to screen 0, it works as intended, but if I have it default to the current screen, it works as intended for screen 0, but for screen 1, if I try to go back to something on screen 0, it moves it to screen 1... is it possible the current screen isn't getting updated?
```python
@lazy.function
def togroup(qtile, g):
    """
        togroup(qtile, g)

    Goes to group `g` on whatever screen it's currently on.
    """
    gg = qtile.groups_map[g.name]
    scr = gg.info()["screen"] or qtile.current_screen.index
    qtile.cmd_to_screen(scr)
    gg.cmd_toscreen()
```

**Reply by elParaguayo** [reply]
Change this line:
``` python
scr = gg.info()["screen"] if  gg.info()["screen"] is not None else qtile.current_screen.index
```

If `gg.info()["screen"]` is 0 that's `False` for the purpose of the test so falls back to `current_screen`. This didn't matter when you wanted to fall back to screen 0 but, now, as you want screen 0 to be valid, you need to explicitly test if there is no screen.

Edit: `is not None` not `is None`

**Reply by ExpandingMan** [reply]
Oh duh, thanks.  My python knowledge is pretty rusty.

**Reply by ExpandingMan** [reply]
Thanks for all your help, one more question, is there a way to move an entire group to a different screen directly?  Or do I have to write a function that goes to the screen first and then switches to the group I was on?  And if the latter, how can I get the *current* group?

**Reply by elParaguayo** [reply]
`lazy.group[group_name].toscreen(screen_index)` (or `lazy.group.toscreen(screen_index)` for current group)

Current group: `qtile.current_group`
